### DISCUSSION TITLE
How to get the server's TLS certificate information?

### DISCUSSION THREAD

**Discussion by sal-limones** [sindresorhus/got#1916] [Q&A] [answered]
```js
https.get(url, { headers, timeout, agent: false }, res => {
  const cert = res.connection.getPeerCertificate()
  const protocol = res.connection.getProtocol()
  const valid = res.socket.authorized
  const error = res.socket.authorizationError
  console.log(cert, protocol, valid, error)
})
``` 

Is there a way to get the same data from Got?

**Comment by wreiske** [comment]
Just noticed this discussion thread after opening an issue here https://github.com/sindresorhus/got/issues/2096

**Comment by sindresorhus** [comment, marked as answer]
You can access the TLS certificate through `response.socket`, but there are some caveats:

```javascript
import got from 'got';

const response = await got('https://example.com');

// Access certificate info
if (response.socket && response.socket.getPeerCertificate) {
	const cert = response.socket.getPeerCertificate();
	const protocol = response.socket.getProtocol();
	const valid = response.socket.authorized;
	const error = response.socket.authorizationError;
	
	console.log(cert, protocol, valid, error);
}
```

**Important caveats:**

1. **Connection reuse**: With agents using keep-alive, `getPeerCertificate()` may return an empty object on subsequent requests to the same host. To avoid this, disable the agent:

```javascript
const response = await got('https://example.com', {
	agent: {
		https: false
	}
});
```

2. **Socket may be closed**: The socket might already be closed when you access it.

**More reliable approach using `checkServerIdentity`:**

```javascript
import got from 'got';

let certificate;

const response = await got('https://example.com', {
	https: {
		checkServerIdentity: (hostname, cert) => {
			certificate = cert;
			return undefined; // Accept the certificate
		}
	}
});

console.log(certificate);
```

This captures the certificate during the TLS handshake, before the connection is established, making it more relible than accessing it after the response.
